ee/desktop/menu/menu.go

```diff
@@ -1,36 +1,150 @@
+	Label       string         `json:"label,omitempty"`
+	Tooltip     string         `json:"tooltip,omitempty"`
+	Disabled    bool           `json:"disabled,omitempty"`
+	NonProdOnly bool           `json:"nonProdOnly,omitempty"`
+	IsSeparator bool           `json:"isSeparator,omitempty"`
+	Action      MenuItemAction `json:"action,omitempty"`
+	Items       []MenuItemData `json:"items,omitempty"`
+}
 
-		// if prod environment, return
-		if hostname == "k2device-preprod.kolide.com" || hostname == "k2device.kolide.com" {
-			return
-		}
+// MenuItemAction encapsulates what action should be performed when a menu item is invoked
+type MenuItemAction struct {
+	URL string `json:"url,omitempty"`
+}
 
-		// in non prod environment
+// MenuBuilder is an interface a menu parser can use to specify how the systray menu is built
+type MenuBuilder interface {
+	SetTooltip(tooltip string)
+	AddMenuItem(label, tooltip string, disabled, nonProdOnly bool, action MenuItemAction, parent any) any
+	AddSeparator()
+}
+
+type menu struct {
+	logger   log.Logger
+	hostname string
+}
+
+func New(logger log.Logger, hostname string) *menu {
+	m := &menu{
+		logger:   logger,
+		hostname: hostname,
+	}
+
+	return m
+}
+
+func (m *menu) Init(buildMenu func()) {
+	systray.Run(buildMenu, nil)
+}
+
+func (m *menu) Build(menu *MenuData) {
+	// Remove all menu items each time we rebuild the menu
+	systray.ResetMenu()
+
+	parseMenuData(menu, m)
+}
+
+func (m *menu) SetTooltip(tooltip string) {
+	systray.SetTooltip(tooltip)
+
+	if m.isProd() {
+		systray.SetTemplateIcon(assets.KolideDesktopIcon, assets.KolideDesktopIcon)
+	} else {
 		systray.SetTemplateIcon(assets.KolideDebugDesktopIcon, assets.KolideDebugDesktopIcon)
-		systray.AddSeparator()
-		systray.AddMenuItem("--- DEBUG ---", "").Disable()
-		systray.AddMenuItem(fmt.Sprintf("Hostname: %s", hostname), "").Disable()
+	}
+}
+
+func (m *menu) AddMenuItem(label, tooltip string, disabled, nonProdOnly bool, action MenuItemAction, parent any) any {
+	if nonProdOnly && m.isProd() {
+		// This is prod environment, but the menu item is for non-prod only
+		return nil
 	}
 
-	systray.Run(onReady, func() {})
+	var item, parentItem *systray.MenuItem
+	parentItem, ok := parent.(*systray.MenuItem)
+	if ok {
+		// If a parent menu item was provided, this is meant to be a sub menu item
+		item = parentItem.AddSubMenuItem(label, tooltip)
+	} else {
+		item = systray.AddMenuItem(label, tooltip)
+	}
+
+	if disabled {
+		item.Disable()
+	}
+
+	// Menu items can have actions associated with them
+	m.addAction(item, action)
+
+	return item
 }
 
-func Shutdown() {
+func (m *menu) AddSeparator() {
+	systray.AddSeparator()
+}
+
+func (m *menu) Shutdown() {
 	systray.Quit()
 }
+
+func (m *menu) isProd() bool {
+	return m.hostname == "k2device-preprod.kolide.com" || m.hostname == "k2device.kolide.com"
+}
+
+func (m *menu) addAction(item *systray.MenuItem, action MenuItemAction) {
+	if action.URL == "" {
+		return
+	}
+
+	go func() {
```

Comment:
I think these goroutines will leak - I'm thinking I have to make done channels which can be signaled when the either the menu items go away, or the systray is exited.

Reply:
done channel makes sense, could signal when you reset he menu

Reply:
Added channels and a cleanup method, verified those goroutines are going away now!